pkg/cert/v1beta1/generate/generate.go

```diff
@@ -0,0 +1,262 @@
+	return cmd
+}
+
+func (o *generateOptions) run(ctx context.Context, kubeClient client.Client) error {
+	o.fullServiceDomain = strings.Join([]string{consts.Service, o.namespace, "svc"}, ".")
+
+	caKeyPair := &certificates{}
+	if err := o.createCACert(caKeyPair); err != nil {
+		return err
+	}
+	keyPair := &certificates{}
+	if err := o.createCert(caKeyPair, keyPair); err != nil {
+		return err
+	}
+
+	if err := o.createWebhookCertSecret(ctx, kubeClient, caKeyPair, keyPair); err != nil {
+		return err
+	}
+	if err := o.injectCert(ctx, kubeClient, caKeyPair); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *generateOptions) createCACert(caKeyPair *certificates) error {
+	now := time.Now()
+	template := &x509.Certificate{
+		SerialNumber: big.NewInt(0),
+		Subject: pkix.Name{
+			CommonName:   consts.CAName,
+			Organization: []string{"katib"},
+		},
+		DNSNames: []string{
+			consts.CAName,
+		},
+		NotBefore:             now,
+		NotAfter:              now.Add(24 * time.Hour * 365 * 10),
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
+
+	var err error
+	caKeyPair.key, err = rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+	der, err := x509.CreateCertificate(rand.Reader, template, template, caKeyPair.key.Public(), caKeyPair.key)
+	if err != nil {
+		return err
+	}
+	if err = caKeyPair.encode(x509.MarshalPKCS1PrivateKey(caKeyPair.key), der); err != nil {
+		return err
+	}
+	if caKeyPair.cert, err = x509.ParseCertificate(der); err != nil {
+		return err
+	}
+	klog.Info("Generate the self-signed CA certificate and private key.")
+	return nil
+}
+
+func (o *generateOptions) createCert(caKeyPair *certificates, keyPair *certificates) error {
+	now := time.Now()
+	template := &x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			CommonName: o.fullServiceDomain,
+		},
+		DNSNames: []string{
+			consts.Service,
+			strings.Join([]string{consts.Service, o.namespace}, "."),
+			o.fullServiceDomain,
+		},
+		NotBefore:             now,
+		NotAfter:              now.Add(24 * time.Hour * 365 * 10),
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: false,
+	}
+
+	var err error
+	keyPair.key, err = rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+	der, err := x509.CreateCertificate(rand.Reader, template, caKeyPair.cert, keyPair.key.Public(), caKeyPair.key)
+	if err != nil {
+		return err
+	}
+	if err = keyPair.encode(x509.MarshalPKCS1PrivateKey(keyPair.key), der); err != nil {
+		return err
+	}
+	klog.Info("Generate public certificate and private key signed with self-singed CA cert and private key.")
+	return nil
+}
+
+// createWebhookCertSecret create Secret embedded ca.key, ca.crt, tls.key and tls.cert
```

Comment:
Please can you also add comments for other functions ?

Reply:
Sure.

Reply:
I added comments to other functions.